### PR TITLE
Visitor can Sign up to app

### PR DESCRIPTION
**Pull request by vick3d**
# Link to feature description(Waffle or Piviotal Tracker)
https://waffle.io/amandagibson/rails_messaging_jan19_team1

Sign-up

# User Story

As a user
In order to access the mailboxer
I want to be able to sign up

# Screenshots

image

image

# Description of PR

Feature and unit testing for sign-up function. 

Fixed functionality of application controller so that it can store the user inputs. 

Tests for happy path and sad path. 
Added a factorybot for User.

### CODE REVIEW

**Review by amandagibson — APPROVED** [review]

**Review by yanqiuzhang — APPROVED** [review]

**Review by amandagibson — APPROVED** [review]

**Review by NiciP — APPROVED** [review]

**Review by NiciP — APPROVED** [review]

**Review by boamatule — APPROVED** [review]

File: features/visitor_is_able_to_sign_up.feature [hunk 1]
```diff
@@ -0,0 +1,42 @@
+Feature: Visitor is able to sign up
+    As a User
+    In order to access the mailboxer
+    I want to be able to Sign Up
+
+Scenario: User can sign up
+
+        Given I visit the page
```
Comment (oliverochman) [inline features/visitor_is_able_to_sign_up.feature]: The first three steps of every scenario are the same, you should move them to a background

File: features/step_definitions/basic_steps.rb [hunk 2]
```diff
@@ -1,3 +1,13 @@
+  Given ("I visit the page") do
+    visit root_path
+  end
+
+  Given("following users exist") do |table|
```
Comment (oliverochman) [inline features/step_definitions/basic_steps.rb]: When are you using this step

File: features/visitor_is_able_to_sign_up.feature [hunk 3]
```diff
@@ -0,0 +1,42 @@
+Feature: Visitor is able to sign up
+    As a User
+    In order to access the mailboxer
+    I want to be able to Sign Up
+
+Scenario: User can sign up
```
Comment (oliverochman) [inline features/visitor_is_able_to_sign_up.feature]: Scenarios should be indented so they line up with the user story

File: features/visitor_is_able_to_sign_up.feature [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        When I click on "Create"
+        Then I should see "Password is too short (minimum is 8 characters)"
+
+
+Scenario: User provides "too many characters for name"
+
+        Given I visit the page
+        When I click on "Sign up"
+        Then I should be redirected to sign up page
+        Then I fill in "Name" with "amandamaegibson"
```
Comment (oliverochman) [inline features/visitor_is_able_to_sign_up.feature]: Dont have two `Then` steps in a row.

File: pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Link to feature description(Waffle or Piviotal Tracker)
```
Comment (oliverochman) [inline pull_request_template.md]: This should be modified for your setup

**Review by oliverochman — CHANGES_REQUESTED** [review]
Need to see more git ponging in future PRs

File: features/step_definitions/basic_steps.rb (hunk 2 above)
Comment (amanda-mae) [reply features/step_definitions/basic_steps.rb]: it's used in user_must_login_to_use_the_app.feature

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,9 @@
+root=true
+
+[*]
+
+charset = utf-8
+indent_style = tab
+indent_size = 2
+trim_trailing_whitespace = true
+insert_final_newline = false
```
Comment (faraznaeem) [inline .editorconfig:9]: What is this?

File: features/visitor_is_able_to_sign_up.feature [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    As a User
+    In order to access the mailboxer
+    I want to be able to Sign Up
+
+Scenario: User can sign up
+
+        Given I visit the page
+        When I click on "Sign up"
+        Then I should be redirected to sign up page
+        Then I fill in "Name" with "Boa"
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature]: `And`

File: features/visitor_is_able_to_sign_up.feature [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        And I fill in "Email" with "dopekids@live.com"
+        And I fill in "Password" with "Wrap009Do"
+        And I fill in "Password confirmation" with "Wrap009Do"
+        When I click on "Create"
+        Then I should see "Welcome! You have signed up successfully."
+
+
+Scenario: User provides insuficient "details for password"
+
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature]: No need for a empty line here

File: features/visitor_is_able_to_sign_up.feature [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+Scenario: User can sign up
+
+        Given I visit the page
+        When I click on "Sign up"
+        Then I should be redirected to sign up page
+        Then I fill in "Name" with "Boa"
+        And I fill in "Email" with "dopekids@live.com"
+        And I fill in "Password" with "Wrap009Do"
+        And I fill in "Password confirmation" with "Wrap009Do"
+        When I click on "Create"
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature]: `And`

**Review by faraznaeem — CHANGES_REQUESTED** [review]

File: .editorconfig (hunk 6 above)
Comment (amanda-mae) [reply .editorconfig:9]: so that we all have the same formatting

Comment (faraznaeem) [reply .editorconfig:9]: Not a blocker, but the feature should only contain what is necessary, this file should be added in a separate PR

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -20,6 +20,8 @@ def conversation
   protected
 
   def configure_permitted_parameters
+    params = [:name, :email, :password, :password_confirmation]
```
Comment (faraznaeem) [inline app/controllers/application_controller.rb]: No need for all of the params, as Devise has built in support for everything except `:name`

File: features/step_definitions/email_inbox_home_page.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+Given("I am logged into my account") do
+expect "Signed in successfully."
+end
+
+
+#Then("I should see my messages.") do
+#	expect(current_path).to eq mailbox_inbox
+#end
```
Comment (faraznaeem) [inline features/step_definitions/email_inbox_home_page.rb]: Remove comments in your code

File: features/step_definitions/email_inbox_home_page.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+Given("I am logged into my account") do
+expect "Signed in successfully."
```
Comment (faraznaeem) [inline features/step_definitions/email_inbox_home_page.rb]: Why is this file in the PR?

File: features/assertions_steps.rb [hunk 13]
```diff
@@ -1,4 +1,7 @@
...
-  end
-  
\ No newline at end of file
+Then("I should see {string}") do |content|
+	expect(page).to have_content content 
+end
+
+Then("I should be redirected to sign up page") do
+	expect(current_path).to eq new_user_registration_path
+end
```
Comment (faraznaeem) [inline features/assertions_steps.rb]: Be consistent with the indentaions

File: features/step_definitions/basic_steps.rb [hunk 14]
```diff
@@ -1,22 +1,25 @@
...
+      end
+  end
+
   When ("I visit the site") do
     visit root_path
   end
-  
+
   When ("I click on {string} link") do |string|
     click_on string
```
Comment (faraznaeem) [inline features/step_definitions/basic_steps.rb]: please dont use string as a variable name

File: features/visitor_is_able_to_sign_up.feature [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+		Scenario: User can sign up
+				And I fill in "Name" with "Boa"
+				And I fill in "Email" with "dopekids@live.com"
+				And I fill in "Password" with "Wrap009Do"
+				And I fill in "Password confirmation" with "Wrap009Do"
+				When I click on "Create"
+				Then I should see "Welcome! You have signed up successfully."
+
+
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature]: Remove extra line

File: features/visitor_is_able_to_sign_up.feature (hunk 8 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: fixed

File: features/visitor_is_able_to_sign_up.feature (hunk 4 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: fixed

File: pull_request_template.md (hunk 5 above)
Comment (amanda-mae) [reply pull_request_template.md]: possibly fixed?

File: features/visitor_is_able_to_sign_up.feature [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+				When I click on "Sign up"
+				Then I should be redirected to sign up page
+
+		Scenario: User can sign up
+				And I fill in "Name" with "Boa"
+				And I fill in "Email" with "dopekids@live.com"
+				And I fill in "Password" with "Wrap009Do"
+				And I fill in "Password confirmation" with "Wrap009Do"
+				When I click on "Create"
+				Then I should see "Welcome! You have signed up successfully."
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature:16]:
```
		Scenario: User can sign up
        When I click on "Sign up"
				And I fill in "Name" with "Boa"
				And I fill in "Email" with "dopekids@live.com"
				And I fill in "Password" with "Wrap009Do"
				And I fill in "Password confirmation" with "Wrap009Do"
				And I click on "Create"
				Then I should see "Welcome! You have signed up successfully."
```

File: features/visitor_is_able_to_sign_up.feature [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+
+		Scenario: User provides insuficient "details for password"
+				And I fill in "Name" with "Boa"
+				And I fill in "Email" with "dopekids@live.com"
+				And I fill in "Password" with "Wrap009"
+				And I fill in "Password confirmation" with "Wrap009"
+				When I click on "Create"
+				Then I should see "Password is too short (minimum is 8 characters)"
```
Comment (faraznaeem) [inline features/visitor_is_able_to_sign_up.feature:24]: Please look at the comment above

File: pull_request_template.md [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+In order to access the mailboxer
+I want to be able to sign up
+
+
+# Screenshots
+
+Add screenshots here for each view/change
+
+# Description of PR
+
```
Comment (faraznaeem) [inline pull_request_template.md:19]: Good

**Review by faraznaeem — CHANGES_REQUESTED** [review]

File: features/visitor_is_able_to_sign_up.feature (hunk 1 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: done

File: features/visitor_is_able_to_sign_up.feature (hunk 3 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: done

File: .editorconfig (hunk 6 above)
Comment (amanda-mae) [reply .editorconfig:9]: noted

File: features/visitor_is_able_to_sign_up.feature (hunk 7 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: done

File: features/visitor_is_able_to_sign_up.feature (hunk 9 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: done

File: app/controllers/application_controller.rb (hunk 10 above)
Comment (amanda-mae) [reply app/controllers/application_controller.rb]: done

File: features/step_definitions/email_inbox_home_page.rb (hunk 11 above)
Comment (amanda-mae) [reply features/step_definitions/email_inbox_home_page.rb]: done

File: features/step_definitions/email_inbox_home_page.rb (hunk 12 above)
Comment (amanda-mae) [reply features/step_definitions/email_inbox_home_page.rb]: should've been deleted. is gone now, sorry.

File: features/assertions_steps.rb (hunk 13 above)
Comment (amanda-mae) [reply features/assertions_steps.rb]: noted

Comment (amanda-mae) [reply features/assertions_steps.rb]: fixed

File: features/step_definitions/basic_steps.rb (hunk 14 above)
Comment (amanda-mae) [reply features/step_definitions/basic_steps.rb]: fixed

File: features/visitor_is_able_to_sign_up.feature (hunk 15 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature]: done

File: features/visitor_is_able_to_sign_up.feature (hunk 16 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature:16]: done

File: features/visitor_is_able_to_sign_up.feature (hunk 17 above)
Comment (amanda-mae) [reply features/visitor_is_able_to_sign_up.feature:24]: done

File: pull_request_template.md (hunk 18 above)
Comment (amanda-mae) [reply pull_request_template.md:19]: thanks

**Comment by amanda-mae** [comment]
changes addressed and fixed.

**Pull request merged**